tensilelite/Tensile/KernelWriterAssembly.py

```diff
@@ -1916,40 +1918,63 @@ def graWorkGroup(self, kernel):
-      module.add(SAddU32(dst=sgpr(wgSerial2), src0=sgpr(wgSerial2), src1=sgpr("WorkGroup0"), comment="wgSerial = wg0 + (wg1 % WGM)*nwg0"))
-
-      module.add(scalarUInt32DivideAndRemainder(qReg=wgmDivisor, dReg="NumWorkGroups1", divReg="WGM", rReg=wgSerial2, tmpVgprRes=tmpVgprRes, wavewidth=kernel["WavefrontSize"], doRemainder=False, comment="WGM"))
-      module.add(SMulI32(dst=sgpr(wgmDivisor2), src0=sgpr("WGM"), src1=sgpr(wgmDivisor), comment="quotient * non-magic divisor"))
-      module.add(SSubU32(dst=sgpr(wgmDivisorMagicNumber), src0=sgpr("NumWorkGroups1"), src1=sgpr(wgmDivisor2), comment="WorkGroup1=remainder"))
-      module.add(SCmpEQU32(src0=sgpr(wgmDivisorMagicNumber), src1=0, comment="remainder == 0 ?"))
-      module.add(SCMovB32(dst=sgpr(wgmDivisorMagicNumber), src=sgpr("WGM"), comment="remainder = WGM if remainder == 0"))
-      module.add(SCmpGeU32(src0=sgpr(blockId2), src1=sgpr(wgmDivisor), comment="blockId >= numFullBlocks ?"))
-      module.add(SCSelectB32(dst=sgpr(wgmDivisor), src0=sgpr(wgmDivisorMagicNumber), src1=sgpr("WGM")))
-
-      # No longer supported for kernel["WorkGroupMapping"] < 0
-      # WorkGroup0 = wgSerial2 / wgmDivisor
-      # WorkGroup1 = wgSerial2 - (wgmDivisor * WorkGroup0)
-      tmpVgpr = self.vgprPool.checkOut(2, "div")
-      tmpVgprRes = RegisterPoolResource(idx=tmpVgpr, size=2)
-      module.add(scalarUInt32DivideAndRemainder(qReg="WorkGroup0", dReg=wgSerial2, divReg=wgmDivisor, rReg="WorkGroup1", tmpVgprRes=tmpVgprRes, wavewidth=kernel["WavefrontSize"]))
-      self.vgprPool.checkIn(tmpVgpr)
-      module.add(SMulI32(dst=sgpr(blockId2), src0=sgpr(blockId2), src1=sgpr("WGM"), comment="blockId * WGM"))
-      module.add(SAddU32(dst=sgpr("WorkGroup1"), src0=sgpr("WorkGroup1"), src1=sgpr(blockId2), comment="wg1 += blockId * WGM"))
+      for wgmType in [False, True]: # Positive/Negative
+        if wgmType:
+          workgroupFirst = "WorkGroup1"
+          workgroupSecond = "WorkGroup0"
+          numWorkgroupsFirst = "NumWorkGroups1"
+          numWorkgroupsSecond = "NumWorkGroups0"
+        else:
+          workgroupFirst = "WorkGroup0"
+          workgroupSecond = "WorkGroup1"
+          numWorkgroupsFirst = "NumWorkGroups0"
+          numWorkgroupsSecond = "NumWorkGroups1"
+
+        if wgmType:
+          module.add(wgmLabelNegative)
+          module.add(SCmpGeI32(src0=sgpr("WGM"), src1=-1, comment="WGM >= -1 ?"))
+          module.add(SCBranchSCC1(labelName=wgmLabel.getLabelName(), comment="branch if WGM >= -1"))
+          module.add(SAbsI32(dst=sgpr("WGM"), src=sgpr("WGM"), comment="abs(WGM)"))
+          module.add(Label("WGMTMP", comment=""))
+        # note this overwrites blockId2+1
+        module.add(scalarUInt32DivideAndRemainder(qReg=blockId2, dReg=workgroupSecond, divReg="WGM", rReg=wgSerial2, tmpVgprRes=tmpVgprRes, wavewidth=kernel["WavefrontSize"], doRemainder=False, comment="WGM"))
```

Comment:
Ideally, we would not like to generate WGM calculation for both positive and negative.
We can unify the code by using some tmp sgpr for first and second.
I think we can overwrite WorkGroup0 and 1 here.
So, what we can do is
   Swap WorkGroup0 and 1 in negative case
   allocate 2 sgpr for numWorkgroupsFirst and second. Store the values properly for positive and negative.

Reply:
I'm not sure if we have enough sgpr. I have to check.

Reply:
Currently some kernels use too much sgpr, so still not able to allocate 2 tmp sgprs.

Reply:
Maybe we can swap  numWorkgroups0 and  numWorkgroups1 and swap back?

Reply:
I was thinking the same way.
I think we can technically do that (the code might look a bit complicated).
We can use xor operation to swap without tmp register.

Reply:
> I was thinking the same way. I think we can technically do that (the code might look a bit complicated). We can use xor operation to swap without tmp register.

But store requires a sgpr to cache the state.

Reply:
We do the abs(WGM) so we can't know if it's negative or not.

Reply:
OK. I do not come up with any other ideas.
We can keep current code for now....

Reply:
Just thinking if we use the same code, we'll need 12 lines to exchange data for negative cases.